File: haskell/algebraic-polynomials/src/Math/Polynomial/Resultant.hs
-- |
-- Module      : Math.Polynomial.Resultant
-- Description : Resultant computation, Lagrange interpolation, and composed polynomials
-- Stability   : experimental
--
-- Canonical implementations of the polynomial resultant (via the Euclidean
-- algorithm), Lagrange interpolation, and composed-sum\/composed-product
-- polynomials. These are generalized to work over any field @k@ with
-- @('Eq' k, 'Fractional' k)@ constraints.
--
-- The composed-sum and composed-product operations are used for computing
-- minimal polynomials of sums and products of algebraic numbers.
module Math.Polynomial.Resultant
  ( polyResultant
  , lagrangeInterpolate
  , composedSum
  , composedProduct
  , negateVar
  , reciprocalPoly
  , substituteXN
  , shiftPoly
  ) where

import Math.Polynomial.Univariate

-- | Resultant of two univariate polynomials via the Euclidean algorithm.
--
-- The resultant \(\text{Res}(f, g)\) is a scalar in the coefficient field
-- that is zero if and only if \(f\) and \(g\) share a common root (over
-- the algebraic closure).
--
-- Complexity: \(O(n \cdot m)\) where \(n = \deg f\) and \(m = \deg g\),
-- via the Euclidean pseudo-remainder sequence.
polyResultant :: (Eq k, Fractional k) => Poly k -> Poly k -> k
polyResultant f g
  | degree f < 0 || degree g < 0 = 0
  | degree g == 0 = case leadCoeff g of
      Just c  -> c ^ degree f
      Nothing -> 0
  | otherwise =
      let (_, r) = divModPoly f g
          df = degree f
          dg = degree g
          lg = case leadCoeff g of Just c -> c; Nothing -> 1
          dr = degree r
          sign = if odd (df * dg) then -1 else 1
      in if degree r < 0
         then 0
         else sign * (lg ^ (df - dr)) * polyResultant g r

-- | Lagrange interpolation: given a list of points \((x_i, y_i)\),
-- compute the unique polynomial of degree \(\leq n\) passing through
-- all \(n + 1\) points.
--
-- __Precondition:__ the \(x_i\) values must be distinct.
--
-- Complexity: \(O(n^2)\) in the number of points.
lagrangeInterpolate :: (Eq k, Fractional k) => [(k, k)] -> Poly k
lagrangeInterpolate points = foldl addPoly zeroPoly terms
  where
    terms = [scalePoly yi (lagrangeBasis xi (map fst points)) | (xi, yi) <- points]
    lagrangeBasis xi xs =
      let others = filter (/= xi) xs
      in foldl mulPoly (constPoly 1)
           [scalePoly (recip (xi - xj)) (mkPoly [-xj, 1]) | xj <- others]

-- | Composed sum polynomial.
--
-- If \(p\) has roots \(\alpha_i\) and \(q\) has roots \(\beta_j\), the
-- composed sum has roots \(\alpha_i + \beta_j\). The result has degree
-- \(\deg p \cdot \deg q\).
--
-- Computed via evaluation and interpolation of
-- \(\text{Res}_y(p(y), q(x - y))\) at sufficiently many points.
composedSum :: (Eq k, Fractional k) => Poly k -> Poly k -> Poly k
composedSum p q =
  let dp = degree p
      dq = degree q
      resultDeg = dp * dq
      points = [fromInteger i | i <- [0..fromIntegral resultDeg]]
      values = [resultantSumAt p q x | x <- points]
  in lagrangeInterpolate (zip points values)

-- | Compute \(\text{Res}_y(p(y), q(x_0 - y))\) for a specific \(x_0\).
resultantSumAt :: (Eq k, Fractional k) => Poly k -> Poly k -> k -> k
resultantSumAt p q x0 =
  let qShifted = substituteLinear q x0 (-1)  -- q(x0 + (-1)*y)
  in polyResultant p qShifted

-- | Substitute \(a + b \cdot y\) for \(x\) in polynomial \(p\): computes
-- \(p(a + b y)\).
substituteLinear :: (Eq k, Fractional k) => Poly k -> k -> k -> Poly k
substituteLinear (Poly []) _ _ = zeroPoly
substituteLinear (Poly cs) a b =
  let binomial = mkPoly [a, b]  -- a + b*y
  in foldr (\c acc -> addPoly (constPoly c) (mulPoly binomial acc)) zeroPoly cs

-- | Composed product polynomial.
--
-- If \(p\) has roots \(\alpha_i\) and \(q\) has roots \(\beta_j\), the
-- result has roots \(\alpha_i \cdot \beta_j\). The result has degree
-- \(\deg p \cdot \deg q\).
--
-- Uses evaluation and interpolation of
-- \(\text{Res}_y(y^{\deg q} \cdot p(x_0/y), q(y))\).
composedProduct :: (Eq k, Fractional k) => Poly k -> Poly k -> Poly k
composedProduct p q =
  let dp = degree p
      dq = degree q
      resultDeg = dp * dq
      points = [fromInteger i | i <- [1..fromIntegral resultDeg + 1]]  -- avoid 0
      values = [productResultantAt p q x | x <- points]
  in lagrangeInterpolate (zip points values)

-- | Compute \(\text{Res}_y(y^n \cdot p(x_0/y), q(y))\) for a specific \(x_0\).
productResultantAt :: (Eq k, Fractional k) => Poly k -> Poly k -> k -> k
productResultantAt p q x0 =
  let pRev = scaledReciprocalAt p x0
  in polyResultant pRev q

-- | Compute \(y^n \cdot p(x_0/y)\) as a polynomial in \(y\).
--
-- If \(p = c_0 + c_1 x + \cdots + c_n x^n\), then
-- \(y^n \cdot p(x_0/y) = c_0 y^n + c_1 x_0 y^{n-1} + \cdots + c_n x_0^n\).
scaledReciprocalAt :: (Eq k, Num k) => Poly k -> k -> Poly k
scaledReciprocalAt (Poly cs) x0 =
  let n = length cs - 1
      newCs = [c * x0 ^ (n - k) | (k, c) <- zip [0..] (reverse cs)]
  in mkPoly newCs

-- | Negate the variable: compute \(p(-x)\) by negating odd-degree coefficients.
--
-- >>> negateVar (mkPoly [1, 2, 3] :: Poly Rational)
-- Poly [1 % 1,(-2) % 1,3 % 1]
negateVar :: (Eq k, Num k) => Poly k -> Poly k
negateVar (Poly cs) = mkPoly $ zipWith (\i c -> if odd (i :: Int) then negate c else c) [0..] cs

-- | Reciprocal polynomial: \(x^n \cdot p(1/x)\), computed by reversing
-- the coefficient list.
--
-- If \(r\) is a root of \(p\), then \(1/r\) is a root of the reciprocal.
reciprocalPoly :: (Eq k, Num k) => Poly k -> Poly k
reciprocalPoly (Poly cs) = mkPoly (reverse cs)

-- | Substitute \(x^n\) for \(x\) in \(p\): compute \(p(x^n)\).
--
-- If \(p = c_0 + c_1 x + c_2 x^2 + \cdots\), then
-- \(p(x^n) = c_0 + c_1 x^n + c_2 x^{2n} + \cdots\).
substituteXN :: (Eq k, Num k) => Int -> Poly k -> Poly k
substituteXN n (Poly cs) =
  let indexed = zip [0 :: Int ..] cs
      maxDeg = (length cs - 1) * n
      result = replicate (maxDeg + 1) 0
      set' xs (i, c) =
        let pos = i * n
        in take pos xs ++ [c] ++ drop (pos + 1) xs
  in mkPoly $ foldl set' result indexed

-- | Shift a polynomial: compute \(f(x + a)\) where \(a\) is in the
-- coefficient field.
--
-- Uses Horner's method for composition with the linear polynomial \(x + a\).
shiftPoly :: (Eq k, Fractional k) => Poly k -> k -> Poly k
shiftPoly (Poly []) _ = zeroPoly
shiftPoly (Poly cs) a =
  -- f(x + a) via Horner
  let xPlusA = mkPoly [a, 1]  -- x + a
  in foldr (\c acc -> addPoly (constPoly c) (mulPoly xPlusA acc)) zeroPoly cs

File: mercury/resultant.m
%---------------------------------------------------------------------------%
% vim: ft=mercury ts=4 sw=4 et
%---------------------------------------------------------------------------%
% File: resultant.m
% Main author: nyc
%
% Resultant computation, Lagrange interpolation, and composed-sum/product
% polynomials over an arbitrary field.
%
%---------------------------------------------------------------------------%

:- module resultant.
:- interface.

:- import_module list.
:- import_module poly.

%---------------------------------------------------------------------------%

    % Resultant of two univariate polynomials via the Euclidean algorithm.
    %
:- func poly_resultant(poly(K), poly(K)) = K <= field(K).

    % Lagrange interpolation: given points {(x_i, y_i)}, compute the unique
    % polynomial of degree <= n passing through all points.
    %
:- func lagrange_interpolate(list({K, K})) = poly(K) <= field(K).

    % Composed sum: if p has roots alpha_i and q has roots beta_j,
    % the result has roots alpha_i + beta_j.
    %
:- func composed_sum(poly(K), poly(K)) = poly(K) <= field(K).

    % Composed product: if p has roots alpha_i and q has roots beta_j,
    % the result has roots alpha_i * beta_j.
    %
:- func composed_product(poly(K), poly(K)) = poly(K) <= field(K).

    % p(-x): negate odd-degree coefficients.
    %
:- func negate_var(poly(K)) = poly(K) <= ring(K).

    % Reciprocal polynomial: x^n * p(1/x), i.e. reverse coefficients.
    %
:- func reciprocal_poly(poly(K)) = poly(K) <= ring(K).

    % Substitute x^n for x in p: p(x^n).
    %
:- func substitute_x_n(int, poly(K)) = poly(K) <= ring(K).

    % Shift a polynomial: f(x + a).
    %
:- func shift_poly(poly(K), K) = poly(K) <= field(K).

%---------------------------------------------------------------------------%
%---------------------------------------------------------------------------%

:- implementation.

:- use_module int.
:- import_module require.
:- import_module string.

%---------------------------------------------------------------------------%

poly_resultant(F, G) = Result :-
    DF = degree(F) : int,
    DG = degree(G) : int,
    ( if int.'<'(DF, 0) then
        Result = ring_zero
    else if int.'<'(DG, 0) then
        Result = ring_zero
    else if DG = (0 : int) then
        ( if lead_coeff(G, C) then
            Result = pow_k(C, DF)
        else
            Result = ring_zero
        )
    else
        div_mod(F, G, _, R),
        ( if lead_coeff(G, LG0) then
            LG = LG0
        else
            LG = ring_one
        ),
        DR = degree(R) : int,
        DFDG = int.'*'(DF, DG),
        Sign = ( if odd(DFDG) then ring_negate(ring_one) else ring_one ),
        ( if int.'<'(DR, 0) then
            Result = ring_zero
        else
            Result = ring_mul(ring_mul(Sign, pow_k(LG, int.'-'(DF, DR))),
                poly_resultant(G, R))
        )
    ).

    % Exponentiation by squaring for ring elements.
    %
:- func pow_k(K, int) = K <= ring(K).

pow_k(X, N) = Result :-
    ( if N = 0 then
        Result = ring_one
    else if int.'>'(N, 0) then
        ( if int.rem(N, 2) = 0 then
            Half = pow_k(X, int.'//'(N, 2)),
            Result = ring_mul(Half, Half)
        else
            Result = ring_mul(X, pow_k(X, int.'-'(N, 1)))
        )
    else
        unexpected($pred, "negative exponent: " ++ string.int_to_string(N))
    ).

:- pred odd(int::in) is semidet.

odd(N) :- int.rem(N, 2) \= 0.

%---------------------------------------------------------------------------%

lagrange_interpolate(Points) = Result :-
    Xs = list.map(fst, Points),
    Terms = list.map(
        ( func({Xi, Yi}) = scale(Yi, lagrange_basis(Xi, Xs)) ),
        Points),
    Result = list.foldl(add, Terms, zero_poly).

:- func lagrange_basis(K, list(K)) = poly(K) <= field(K).

lagrange_basis(Xi, Xs) = Result :-
    Others = list.filter(( pred(X::in) is semidet :- X \= Xi ), Xs),
    Result = list.foldl(
        ( func(Xj, Acc) = mul(Acc, scale(
            field_div(ring_one, ring_sub(Xi, Xj)),
            mk_poly([ring_negate(Xj), ring_one]))) ),
        Others, const_poly(ring_one)).

%---------------------------------------------------------------------------%

composed_sum(P, Q) = Result :-
    DP = degree(P) : int,
    DQ = degree(Q) : int,
    ResultDeg = int.'*'(DP, DQ),
    Points = make_points(0, ResultDeg),
    Values = list.map(
        ( func(X) = resultant_sum_at(P, Q, X) ),
        Points),
    Result = lagrange_interpolate(zip_pairs(Points, Values)).

:- func resultant_sum_at(poly(K), poly(K), K) = K <= field(K).

resultant_sum_at(P, Q, X0) = poly_resultant(P, QShifted) :-
    QShifted = substitute_linear(Q, X0, ring_negate(ring_one)).

    % substitute_linear(P, A, B) = P(A + B*y).
    %
:- func substitute_linear(poly(K), K, K) = poly(K) <= field(K).

substitute_linear(poly(Cs), A, B) = Result :-
    Binomial = mk_poly([A, B]),
    list.reverse(Cs, RevCs),
    list.foldl(
        ( pred(C::in, Acc0::in, Acc1::out) is det :-
            Acc1 = add(const_poly(C), mul(Binomial, Acc0))
        ), RevCs, zero_poly, Result).

%---------------------------------------------------------------------------%

composed_product(P, Q) = Result :-
    DP = degree(P) : int,
    DQ = degree(Q) : int,
    ResultDeg = int.'*'(DP, DQ),
    Points = make_points(1, int.'+'(ResultDeg, 1)),
    Values = list.map(
        ( func(X) = product_resultant_at(P, Q, X) ),
        Points),
    Result = lagrange_interpolate(zip_pairs(Points, Values)).

:- func product_resultant_at(poly(K), poly(K), K) = K <= field(K).

product_resultant_at(P, Q, X0) = poly_resultant(PRev, Q) :-
    PRev = scaled_reciprocal_at(P, X0).

    % y^n * p(x0/y) as a polynomial in y.
    %
:- func scaled_reciprocal_at(poly(K), K) = poly(K) <= ring(K).

scaled_reciprocal_at(poly(Cs), X0) = mk_poly(NewCs) :-
    N = int.'-'(list.length(Cs), 1),
    list.reverse(Cs, RevCs),
    NewCs = list.map_corresponding(
        ( func(Idx, C) = ring_mul(C, pow_k(X0, int.'-'(N, Idx))) ),
        make_int_list(0, N), RevCs).

%---------------------------------------------------------------------------%

negate_var(poly(Cs)) = mk_poly(NegCs) :-
    negate_odd(Cs, (0 : int), NegCs).

:- pred negate_odd(list(K)::in, int::in, list(K)::out) is det <= ring(K).

negate_odd([], _, []).
negate_odd([C | Cs], I, [C1 | Rest]) :-
    ( if odd(I) then C1 = ring_negate(C) else C1 = C ),
    negate_odd(Cs, int.'+'(I, 1), Rest).

reciprocal_poly(poly(Cs)) = mk_poly(list.reverse(Cs)).

substitute_x_n(N, poly(Cs)) = mk_poly(Result) :-
    ( if Cs = [] then
        Result = []
    else
        MaxDeg = int.'*'(int.'-'(list.length(Cs), 1), N),
        Zeros = list.duplicate(int.'+'(MaxDeg, 1), ring_zero),
        set_spaced(Cs, 0, N, Zeros, Result)
    ).

:- pred set_spaced(list(K)::in, int::in, int::in,
    list(K)::in, list(K)::out) is det <= ring(K).

set_spaced([], _, _, !Acc).
set_spaced([C | Cs], Pos, Stride, !Acc) :-
    list.det_replace_nth(!.Acc, int.'+'(Pos, 1), C, !:Acc),
    set_spaced(Cs, int.'+'(Pos, Stride), Stride, !Acc).

shift_poly(poly(Cs), A) = Result :-
    ( if Cs = [] then
        Result = zero_poly
    else
        XPlusA = mk_poly([A, ring_one]),
        list.reverse(Cs, RevCs),
        list.foldl(
            ( pred(C::in, Acc0::in, Acc1::out) is det :-
                Acc1 = add(const_poly(C), mul(XPlusA, Acc0))
            ), RevCs, zero_poly, Result)
    ).

%---------------------------------------------------------------------------%
% Utility
%---------------------------------------------------------------------------%

    % make_points(Start, End) = list of ring_from_int(Start..End).
    %
:- func make_points(int, int) = list(K) <= ring(K).

make_points(Start, End) = Result :-
    ( if int.'>'(Start, End) then
        Result = []
    else
        Result = [ring_from_int(Start) | make_points(int.'+'(Start, 1), End)]
    ).

:- func make_int_list(int, int) = list(int).

make_int_list(Start, End) =
    ( if int.'>'(Start, End) then
        []
    else
        [Start | make_int_list(int.'+'(Start, 1), End)]
    ).

:- func zip_pairs(list(K), list(K)) = list({K, K}).

zip_pairs(As, Bs) =
    ( if As = [A | At], Bs = [B | Bt] then
        [{A, B} | zip_pairs(At, Bt)]
    else
        []
    ).

:- func fst({A, B}) = A.
fst({X, _}) = X.

%---------------------------------------------------------------------------%
:- end_module resultant.
%---------------------------------------------------------------------------%
